smart-connector-rest-dist/src/test/java/eu/knowledge/engine/rest/api/TestAskAnswerReactWithGapsEnabled.java

```diff
@@ -0,0 +1,193 @@
+					test.expectStatus(200);
+	
+					// build the body to react to the request: add handle request ID and dummy data bindingset
+					JsonObjectBuilder builder = Json.createObjectBuilder();
+					JsonReader jp = Json.createReader(new StringReader(test.getBody()));
+					JsonObject jo = jp.readObject();
+					int handleRequestId = jo.getInt("handleRequestId");
+					builder.add("handleRequestId", handleRequestId);
+					// for now simply add an empty bindingset as a result
+					builder.add("bindingSet", JsonObject.EMPTY_JSON_ARRAY);
+					JsonObject jo2 = builder.build();
+					String body = jo2.toString();
+					System.out.println("Handle a react to a request with body: " + body);
+
+					// fire the POST handle to execute the react
+					var test2 = new HttpTester(new URL(url.toString() + "/sc/handle"), "POST", body,
+							Map.of("Knowledge-Base-Id", reactKBId, "Knowledge-Interaction-Id", reactKIId, "Content-Type",
+									"application/json", "Accept", "*/*"));
+					test2.expectStatus(200);
+
+				} catch (MalformedURLException e) {
+					fail();
+				}
+			}
+		});
+		reactingSc.start();
+		
+		// register the AskKB
+		HttpTester registerKb = new HttpTester(new URL(url + "/sc"), "POST",
+				"{\"knowledgeBaseId\": \"https://www.tno.nl/example/relationAsker\", \"knowledgeBaseName\": \"RelationAsker\", \"knowledgeBaseDescription\": \"A KB that asks for relations between people\", \"reasonerEnabled\" : true}",
+				Map.of("Content-Type", "application/json", "Accept", "*/*"));
+		registerKb.expectStatus(200);
+
+		// register the AskKI
+		HttpTester registerKiWithoutGapsEnabled = new HttpTester(new URL(url + "/sc/ki"), "POST", """
+				{
+				 "knowledgeInteractionType": "AskKnowledgeInteraction",
+				 "knowledgeInteractionName": "askRelations",
+				 "graphPattern": "?a <http://example.org/isRelatedTo> ?b . ?a <http://example.org/isFatherOf> ?c .",
+				 "knowledgeGapsEnabled": true
+				}
+				""", Map.of("Knowledge-Base-Id", "https://www.tno.nl/example/relationAsker", "Content-Type", "application/json", "Accept",
+						"*/*"));
+		registerKiWithoutGapsEnabled.expectStatus(200);
+
+		// fire the ask KI
+		HttpTester askKiWithoutGapsEnabled = new HttpTester(new URL(url + "/sc/ask"), "POST",
+				"{\"recipientSelector\": {\"knowledgeBases\": []}, \"bindingSet\": []}",
+				Map.of("Knowledge-Base-Id", "https://www.tno.nl/example/relationAsker", "Knowledge-Interaction-Id", "https://www.tno.nl/example/relationAsker/interaction/askRelations",
+						"Content-Type", "application/json", "Accept",
+						"*/*"));
+		var result = askKiWithoutGapsEnabled.getBody();
+		System.out.println("Result is:" +result);
+		assertTrue(result.contains("\"knowledgeGaps\":[[\"?a <http://example.org/isFatherOf> ?c\",\"?a <http://example.org/liveInTheSameHouse> ?b\"]]"));		
+		assertTrue(result.contains("\"bindingSet\":[]"));		
+		
+	}
+
+	@AfterAll
+	public void cleanUp() {
+		rsh.cleanUp();
+	}
```

Comment:
Hey @JackJackie, all of these tests succeed individually and during a Maven build, but some of them fail when I run them directly from my Eclipse IDE environment. Apparently, in that scenario all these tests run after one another in the same JVM and thus using the same KE Runtime. And since all tests (also the older ones) do fully not clean up their KBs before finishing (they only stop the REST Server, but the KE Runtime stays active and so the registered KBs remain active), if the next test uses the same Knowledge Base IDs (which the above test does with for example `https://www.tno.nl/example/relationAsker`) these will already be registered and give an http status code 400 with the following error:

`{"messageType":"error","message":"That knowledge base ID is already in use."}`

I contributed to this branch and created a generalized solution for the above problem by introducing the `TestUtil#unregisterAllKBs(String)` class and method that remove all registered KBs from a REST Server automatically. I call this method at the end of every REST API Unit Test and this indeed seems to fix it.